transport/errors.go

```diff
@@ -61,44 +77,106 @@ func AsHandlerError(service, procedure string, err error) error {
 	}
 }
 
-// isHandlerError may be mixed into types which are valid HandlerErrors.
-type isHandlerError struct{}
-
-func (isHandlerError) handlerError() {}
-
 //////////////////////////////////////////////////////////////////////////////
 
 // BadRequestError is a failure to process a request because the request was
 // invalid.
-type BadRequestError struct {
-	isHandlerError
+type BadRequestError interface {
+	error
+
+	badRequestError()
+}
 
+type localBadRequestError struct {
 	Reason error
 }
 
-func (e BadRequestError) Error() string {
+var _ BadRequestError = localBadRequestError{}
+var _ HandlerError = localBadRequestError{}
+
+// LocalBadRequestError wraps the given error into a BadRequestError.
+//
+// It represents a local failure while processing an invalid request.
+func LocalBadRequestError(err error) HandlerError {
+	return localBadRequestError{Reason: err}
+}
+
+func (localBadRequestError) handlerError()    {}
+func (localBadRequestError) badRequestError() {}
+
+func (e localBadRequestError) Error() string {
 	return "BadRequest: " + e.Reason.Error()
 }
 
-// AsHandlerError on a BadRequestError returns the error as-is.
-func (e BadRequestError) AsHandlerError() HandlerError { return e }
+type remoteBadRequestError struct {
+	Message string
+}
+
+var _ BadRequestError = remoteBadRequestError{}
+
+// RemoteBadRequestError builds a new BadRequestError with the given message.
+//
+// It represents a BadRequest failure from a remote service.
+func RemoteBadRequestError(message string) BadRequestError {
+	return remoteBadRequestError{Message: message}
+}
+
+func (remoteBadRequestError) badRequestError() {}
+
+func (e remoteBadRequestError) Error() string {
+	return e.Message
+}
+
+//////////////////////////////////////////////////////////////////////////////
 
 // UnexpectedError is a failure to process a request for an unexpected reason.
-type UnexpectedError struct {
-	isHandlerError
+type UnexpectedError interface {
```

Comment:
Can you update this comment to add an easy-to-understand example of when users might run into `UnexpectedError`?